grpc.go

```diff
@@ -188,9 +188,17 @@ func (a *agentGRPC) getContainer(cid string) (*container, error) {
 	return ctr, nil
 }
 
-// Shared function between StartContainer and ExecProcess, because those expect
-// a process to be run.
-func (a *agentGRPC) execProcess(ctr *container, proc *process) (err error) {
+// Shared function between CreateContainer and ExecProcess, because those expect
+// a process to be run, and the console to be properly setup.
```

Comment:
I see where you want to go with that one, but it feels like the `execProcess` function is doing exec and non exec operations now.
What you're trying to factorize here is the process post start phase from `execProcess()`. IMHO the following would make the code more readable:

1. Factorize the process post start code:

```Go
func (a *agentGRPC) postStartProcess(ctr *container, proc *process) (err error) {
	// Get process PID
	pid, err := proc.process.Pid()
	if err != nil {
		return err
	}

	proc.exitCodeCh = make(chan int, 1)

	// Create process channel to allow WaitProcess to wait on it.
	// This channel is buffered so that reaper.reap() will not
	// block until WaitProcess listen onto this channel.
	a.sandbox.subreaper.setExitCodeCh(pid, proc.exitCodeCh)

	// Setup terminal if enabled.
	if proc.consoleSock != nil {
		termMaster, err := utils.RecvFd(proc.consoleSock)
		if err != nil {
			return err
		}

		if err := setConsoleCarriageReturn(int(termMaster.Fd())); err != nil {
			return err
		}

		proc.termMaster = termMaster
	}

        ctr.setProcess(proc)

	return nil
}
```

2. Reduce `execProcess()` to
```Go
func (a *agentGRPC) execProcess(ctr *container, proc *process) (err error) {
	// This lock is very important to avoid any race with reaper.reap().
	// Indeed, if we don't lock this here, we could potentially get the
	// SIGCHLD signal before the channel has been created, meaning we will
	// miss the opportunity to get the exit code, leading WaitProcess() to
	// wait forever on the new channel.
	// This lock has to be taken before we run the new process.
	a.sandbox.subreaper.RLock()
	defer a.sandbox.subreaper.RUnlock()

	if proc == nil {
		err = ctr.container.Exec()
	} else {
		err = ctr.container.Run(&(proc.process))
	}

        return err
}
```

3. Call `postStartProcess()` after `execProcess()` in both the `start` and `run` cases.
4. Call `postStartProcess()` after  `container.container.Start()` for the `create` case.

Reply:
Several things here. I agree this would be a bit cleaner but we cannot apply this as you've described it, otherwise we won't hold the lock when setting the exit code channel, which is pretty important. We cannot release the lock after we've started the process, and until we have set the exit code channel. We need this lock both to prevent the subreaper to reap the first process started by libcontainer and second to prevent the subreaper to reap a terminated process before the exit code channel has been registered.
I could maybe split differently, by retrieving the pid and setting the exit code channel from `execProcess()`, and do the rest from `postStartProcess()`.
Another thing, I don't want to call into `postStartProcess()` for the `start` case because the only need to do is call `Exec()`, nothing else.

Reply:
Right, I did not look at those details, but the idea was that `execProcess()` should only be used for starting and exec'ing. So if you can find the right split to do so, it would be cleaner indeed.